buffer/src/main/java/io/netty/buffer/WrappedUnpooledUnsafeDirectByteBuf.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright 2016 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer;
+
+import io.netty.util.internal.PlatformDependent;
+
+import java.nio.ByteBuffer;
+
+final class WrappedUnpooledUnsafeDirectByteBuf extends UnpooledUnsafeDirectByteBuf {
+
+    WrappedUnpooledUnsafeDirectByteBuf(ByteBufAllocator alloc, long memoryAddress, int size, boolean doFree) {
+        super(alloc, PlatformDependent.directBuffer(memoryAddress, size), size, doFree);
+    }
+
+    @Override
+    protected void freeDirect(ByteBuffer buffer) {
+        PlatformDependent.freeMemory(memoryAddress());
```

Comment:
`memoryAddress()` calls `ensureAccessible()` but at that point the buffer's `refCnt` is already 0 and throws `IllegalReferenceCountException`. 

Invocation order is:
```java
Exception in thread "Thread-0" io.netty.util.IllegalReferenceCountException: refCnt: 0
	at io.netty.buffer.AbstractByteBuf.ensureAccessible(AbstractByteBuf.java:1425)
	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.memoryAddress(UnpooledUnsafeDirectByteBuf.java:218)
	at io.netty.buffer.WrappedUnpooledUnsafeDirectByteBuf.freeDirect(WrappedUnpooledUnsafeDirectByteBuf.java:30)
	at io.netty.buffer.UnpooledUnsafeDirectByteBuf.deallocate(UnpooledUnsafeDirectByteBuf.java:490)
	at io.netty.buffer.AbstractReferenceCountedByteBuf.release(AbstractReferenceCountedByteBuf.java:138)
```

One solution is to mark `UnpooledUnsafeDirectByteBuf::memoryAddress` as protected and reference it here instead of calling `memoryAddress()`.